File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.cpp
#include <string>
#include <vector>
#include <map>
#include <iostream>
#include <algorithm>

using namespace std;

vector<string> solution(vector<string> play, vector<string> callings) {
    map<string, int> mymap;
    int call_idx;
    string front_name;
    
    for ( int i = 0; i < play.size(); i++) {
        string name = play[i];
        mymap[name] = i; 
        // map에 name을 인덱스, i를 val로 저장
    }
    
    for(auto c : callings) {
        call_idx = mymap[c]; // c의 위치
        front_name = play[call_idx - 1]; // c 앞에 있는 애의 이름
        swap(play[call_idx], play[call_idx - 1]); // 실제 play 배열도 업데이트
        
        mymap[c]--, mymap[front_name]++;
        
    }
    
    
    return play;
}

File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.py
from collections import deque

def solution(play, callings):
    answer = []
    dic = {}
    
    for idx, p in enumerate(play):
        dic[p] = idx
    
    for call in callings:
        # dic[call] = idx 나타냄
        # play = 실제 순서(정답)
        # dic = 이름 -> 인덱스
        
        call_pos = dic[call] #call의 원래 위치 저장
        name = play[call_pos-1] #call 앞에 있는 애의 이름
        
        play[dic[call]-1], play[dic[call]] = play[dic[call]], play[dic[call]-1] # 실제 추월 상황 반영
        
        dic[call] -= 1 # 순위 저장도 업데이트
        dic[name] += 1      

    answer = play
    return answer